packaging/build_rpm.py

```diff
@@ -0,0 +1,375 @@
+            'chown', '-R', 'mockbuild', '/opt'
+            ])
+    # Build the RPM
+    check_call([
+            MOCK, src_rpm,
+            '--no-clean',
+            ])
+
+    # Extract the final .rpm file name.
+    for f in listdir(RESULT_DIR):
+        if f.endswith('.x86_64.rpm') or f.endswith('.noarch.rpm'):
+            final_rpm = path_join(RESULT_DIR, f)
+            break
+    else:
+        raise RuntimeError('final rpm not found')
+
+    return final_rpm
+
+
+def get_rpmbuild_defines():
+    """
+    Return a dictionary of macros from `version_info`
+    """
+    defines = {}
+    with open(path_join(SCRIPT_DIR, 'version_info')) as f:
+        for line in f:
+            line = line.strip()
+            if line and not line.startswith('#'):
+                k, v = line.split(' ', 1)
+                defines[k] = v
+
+    return defines
+
+
+def render_spec_file(spec_file, defines):
+    """
+    Apply the defines to the spec file. We don't use `--define` because
+    `mock`'s `--rpmbuild-opts` doesn't apply to `--buildsrpm`
+
+    Returns the path to the rendered spec file
+    """
+    with open(spec_file) as f:
+        spec_content = f.read()
+
+    for k, v in defines.items():
+        spec_content = spec_content.replace('%{' + k + '}', v)
+
+    spec_file_out = spec_file.replace('.spec', '.rendered.spec')
+    with open(spec_file_out, 'w') as f:
+        f.write(spec_content)
+
+    return spec_file_out
+
+
+def run_vagrant(cmd, *args, **kwargs):
+    """Run command on Vagrant box"""
+    func = kwargs.pop('_func', check_call)
+
+    logger.info('running <{cmd}> on vagrant builder'.format(cmd=cmd))
+    return func(
+            ['vagrant', 'ssh', 'builder', '-c', cmd],
+            *args, **kwargs)
+
+
+def get_dependencies(spec_file):
+    """get the list of dependencies"""
+    dependencies_file = DEPENDENCIES_FILE.format(spec_file=spec_file)
+
+    try:
+        with open(dependencies_file) as f:
+            dependencies = f.readlines()
+    except IOError:
+        logger.info(
+                'no dependencies file found. skipping dependencies install '
+                '(this is only required for internal cloudify dependencies)')
+        dependencies = []
+
+    return dependencies
+
+
+def install_dependencies(spec_file, source):
+    """
+    Check for and install any internal dependencies.
+
+    Place dependencies, one package name per line, in a file called
+    '<spec_file_name>.dependencies'
+
+    e.g.:
+        $ cat cloudify-premium/packages/cloudify-premium.spec.dependencies
+        cloudify-rest-service
+    """
+    dependencies = get_dependencies(spec_file)
+
+    if not dependencies:
+        return
+
+    logger.info(
+            'installing dependencies from <{file}>. RPMs dir: {dir}'.format(
+                file=DEPENDENCIES_FILE.format(spec_file=spec_file),
+                dir=RPMS_DIR,
+                ))
+
+    rpms = {}
+    for f in listdir(RPMS_DIR):
```

Comment:
I'm not entirely sure what each of the two for loops does. The first one goes over the current folder and looking for rpms to install, and the second loop looks for dependencies from the .spec.dependencies file?
Maybe add some more comments?

Reply:
Why don't we just go ahead and download all the necessary dependencies in advance and put them all in the same folder, and just install everything in there?

Reply:
The first loop finds the RPMs and extracts the package names from them (i.e. `cloudfy-rest-service` from `cloudify-rest-service-4.3.blahblah.rpm`)
The second loop installs the RPMs it needs. Because the RPMs directory is the build script directory, there will be multiple RPMs sitting here on local runs, and we don't want to install all of them
(actually this should probably be extended to install the correct version if there are multiple with the same name)

Reply:
We don't download all in advance because some of what we want to install should come from another one of these spec files.
Doing it this way means I can do a local build of `-rest-service` and then the RPM I need is saved ready for a `-premium` build, no upload/download required. In Jenkins we achieve the same by copying artifacts as you know